File: 백준/Gold/1007. 벡터 매칭/벡터 매칭.py
import sys
import math
from itertools import combinations

input = sys.stdin.readline

T = int(input())

for _ in range(T):
    N = int(input())
    points = [tuple(map(int, input().split())) for _ in range(N)]

    total_x = sum(x for x, y in points)
    total_y = sum(y for x, y in points)

    ans = float('inf')

    # N/2개 선택
    for comb in combinations(range(N), N // 2):
        sx = sy = 0
        for i in comb:
            sx += points[i][0]
            sy += points[i][1]

        # 벡터 합 = (+그룹) - (-그룹)
        vx = 2 * sx - total_x
        vy = 2 * sy - total_y

        length = math.sqrt(vx * vx + vy * vy)
        ans = min(ans, length)

    print(f"{ans:.12f}")

File: 백준/Gold/1007. 벡터 매칭/벡터 매칭.java
import java.io.*;
import java.util.*;

public class Main {

    static int N;
    static long[] x, y;
    static long totalX, totalY;
    static double answer;

    static void dfs(int idx, int cnt, long sx, long sy) {
        if (cnt == N / 2) {
            long vx = 2 * sx - totalX;
            long vy = 2 * sy - totalY;
            double len = Math.sqrt((double) vx * vx + (double) vy * vy);
            answer = Math.min(answer, len);
            return;
        }
        if (idx == N) return;
        if (cnt + (N - idx) < N / 2) return;

        // 선택
        dfs(idx + 1, cnt + 1, sx + x[idx], sy + y[idx]);
        // 미선택
        dfs(idx + 1, cnt, sx, sy);
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int T = Integer.parseInt(br.readLine());

        while (T-- > 0) {
            N = Integer.parseInt(br.readLine());
            x = new long[N];
            y = new long[N];

            totalX = 0;
            totalY = 0;

            for (int i = 0; i < N; i++) {
                st = new StringTokenizer(br.readLine());
                x[i] = Long.parseLong(st.nextToken());
                y[i] = Long.parseLong(st.nextToken());
                totalX += x[i];
                totalY += y[i];
            }

            answer = Double.MAX_VALUE;
            dfs(0, 0, 0, 0);

            System.out.printf("%.12f%n", answer);
        }
    }
}
